proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+offered by `Cabal` and/or build tools to packages using `Hooks`.  There is a
+tension here, because we do not want package maintainers to be over-burdened
+with continual changes to support newer `Cabal` versions, but neither do we want
+`Cabal` maintainers to be over-burdened by the costs of providing backwards
+compatibility.
+
+Since `Hooks` allows the build system to be extended, arguably packages using
+this feature necessarily depend on the version of the build system (and hence
+the `Cabal` version). Thus it may be reasonable to expect package authors to
+declare the range of `Cabal` versions for which their package works (in
+`setup-depends`) and require it to be updated to support newer versions of build
+tools.
+
+While there are still questions to be resolved here about the compatibility
+guarantees that can be offered by build tools, the `Hooks` build-type makes the
+situation no worse than with `Custom` (because a shim `Setup.hs` can still
+always be used to compile `SetupHooks.hs` using an older `Cabal`), but gives
+significantly more options to the build tool.  Where serialisation of hook
+inputs/outputs is used, the serialisation format can be controlled by the build
+tool, and is not necessarily fixed by `Cabal`.  Indeed, we could imagine a build
+tool being compiled against multiple `Cabal` versions.
+
+Note that the version compatibility problem exists in `cabal-install` already:
+even where communication happens via the `Setup.hs` command line interface,
+there is already a need for `cabal-install` to adapt to the command-line flags
+that are supported by the version of `Cabal` in use (see
+e.g. [`filterConfigureFlags`](https://hackage.haskell.org/package/cabal-install-3.10.2.1/docs/src/Distribution.Client.Setup.html#filterConfigureFlags)).
+Once this proposal removes the need for `cabal-install` to go through the
+`Setup.hs` interface, there is a potential for significant reduction in
+complexity here.
+
+
+### Recompilation avoidance
+
+We could imagine allowing hooks to return dependency information to support
+Cabal's recompilation checker in deciding whether it needs to rerun the hooks or
+not. This is particularly relevant when building documentation, as ideally one
+doesn't want to have to run the hooks twice in the workflow `cabal build &&
+cabal haddock`.
+
+A natural way to do this that fits with Cabal's current design would be to have
+a hook that returns the files consulted by each phase, then rerun the phase only
+if those files change.  For example, alongside
+```haskell
+preConfPackageHook :: Maybe (PreConfPackageInputs -> IO PreConfPackageOutputs)
+```
+we could have
+```haskell
+confPackageRecompile :: Maybe (PreConfPackageInputs -> IO Dependencies)
+```
+where `Dependencies` is some representation of possible dependencies
+(e.g. files, file path globs, environment variables, etc.) that may be consulted
+by `preConfPackageHook` and should cause recompilation if they change.
```

Comment:
Are you saying hooks should carefully not write files if nothing changed? That seems like work people will not do. I think hooks should always write files / not worry about recompilation avoidance, and the host build system should do what it needs to do (adjust timestamps, just relying on content addresses).

Dynamically specifying what dependencies one ended up reading is very error prone. Packages will not do this correctly. We should not have that be part of the interface. The only thing which is robust and maintainable is you pre-declare what you care about. Ideally we eventually have some sort of sandboxing so attempting to read anything other than those predeclared dependencies will fail.

We've seen that incremental builds with Hadrian are completely sketchy, and thus no one is even attempting to do incremental builds in CI. I thus consider the non-correct-by-construction ways of doing dependency analysis a complete **social failure**, and we should not attempt them any further.

Reply:
A "stricter" way to do this would be to run the hook in a restricted monad with four functions made available -- one for reading files, one for writing files, one for listing directories, and one for executing processes. The _caller_ could choose these functions when running the monad -- and e.g. the read file one could record which files were read for future cache invalidation checking, the write files one could check the directories written to against some sort of restricted list of "safe" ones if so desired (or at least record what was written for future `clean` purposes), and soforth. Initially, the functions could be straight passthrus to the standard ones, but over time different tools (including cabal) could augment them in useful ways. This would also make it harder than not for users to fail to record deps.

I recognize this isn't perfect -- after all, what if you call an executable, and the problem is the files the _executable_ reads (but this can be done with custom syscall tracing in the invoke process command), etc. But it at least gives us some power to leverage going forward.